### PR TITLE
Fix ARM integration tests

### PR DESCRIPTION
**Pull request by jefferbrecht**
## Description
This fixes the failing tests from #1306:

* Failing Ops Agent tests:
  * ~TestPrometheusMetricsWithJSONExporter/debian-11-arm64~ (fixed)
* Failing 3P tests:
  * ~TestThirdPartyApps/debian-11-arm64/mysql~ (skipped in favour of mariadb)
  * ~TestThirdPartyApps/debian-11-arm64/mysql5.7~ (skipped)
  * ~TestThirdPartyApps/debian-11-arm64/mongodb3.6~ (skipped)
  * ~TestThirdPartyApps/debian-11-arm64/cassandra~ (fixed)
  * ~TestThirdPartyApps/debian-11-arm64/mongodb~ (fixed)
  * ~TestThirdPartyApps/debian-11-arm64/couchbase~ (skipped)
  * ~TestThirdPartyApps/debian-11-arm64/oracledb~ (skipped)
* WIP:
  * TestThirdPartyApps/debian-11-arm64/mariadb

## Related issue
b/274913389

## How has this been tested?
Will let presubmits run.

## Checklist:
- Unit tests
  - [x] Unit tests do not apply.
  - [ ] Unit tests have been added/modified and passed for this PR.
- Integration tests
  - [ ] Integration tests do not apply.
  - [x] Integration tests have been added/modified and passed for this PR.
- Documentation
  - [x] This PR introduces no user visible changes.
  - [ ] This PR introduces user visible changes and the corresponding documentation change has been made.
- Minor version bump
  - [x] This PR introduces no new features.
  - [ ] This PR introduces new features, and there is a separate PR to bump the [minor version](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/VERSION) since the last [release](https://github.com/GoogleCloudPlatform/ops-agent/releases) already.
  - [ ] This PR bumps the version.

### CODE REVIEW

**Comment by quentinmit** [comment]
Debian 11 ships native MySQL on ARM, why aren't we testing that?

https://packages.debian.org/bullseye/mariadb-server

**Comment by jefferbrecht** [comment]
> Debian 11 ships native MySQL on ARM, why aren't we testing that?
> 
> https://packages.debian.org/bullseye/mariadb-server

I definitely forgot about mariadb and I can work on adding this.

File: integration_test/third_party_apps_data/applications/cassandra/debian_ubuntu/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/cassandra/debian_ubuntu/install:14]: This isn't supposed to be how Debian handles conflicting binary names... each package that provides a binary (like `java`) has a priority, and every time you install or remove one of those packages it chooses the highest-priority version out of the currently installed ones.

Are you sure it actually has a transitive dependency on Java 17, and not just a dependency on a JRE that apt is resolving to Java 17 by default? If it's the latter, you can just specify both package names on a single `apt install`.

File: integration_test/third_party_apps_data/applications/couchbase/metadata.yaml
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/couchbase/metadata.yaml:201]: This comment is no longer correct; Debian 11 is now supported.

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install:5]: FWIW `apt-key add` is deprecated and will be removed in Debian 12. You're supposed to just place the file in `/etc/apt/trusted.gpg.d/` directly instead (i.e. pass `--output-document=/etc/apt/trusted.gpg.d/mongodb.asc` to `wget`).

File: integration_test/third_party_apps_data/applications/cassandra/debian_ubuntu/install
Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/cassandra/debian_ubuntu/install:14]: I'm not a fan of this code either, but putting them in a single `apt install` simply doesn't work. Cassandra just picks whatever JRE is on PATH, which ends up being Java 17 when installed this way, which it doesn't support:

```
$ sudo apt install -y openjdk-11-jdk cassandra
$ /usr/sbin/cassandra
OpenJDK 64-Bit Server VM warning: Option UseBiasedLocking was deprecated in version 15.0 and will likely be removed in a future release.
Unrecognized VM option 'UseConcMarkSweepGC'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
$ java --version
openjdk 17.0.7 2023-04-18
OpenJDK Runtime Environment (build 17.0.7+7-Debian-1deb11u1)
OpenJDK 64-Bit Server VM (build 17.0.7+7-Debian-1deb11u1, mixed mode, sharing)
```

If I do `apt depends cassandra` then I get:

```
$ sudo apt depends cassandra                                                                                                                       
cassandra                                                                                                                                                                   
 |Depends: <openjdk-8-jre-headless>                                                                                                                                         
 |Depends: <java8-runtime>                                                                                                                                                  
    default-jre                                                                                                                                                             
    openjdk-11-jre                                                                                                                                                          
    openjdk-17-jre                                                                                                                                                          
 |Depends: openjdk-11-jre-headless
  Depends: <java11-runtime>
    default-jre
    openjdk-11-jre
    openjdk-17-jre
...
```

Now, if I do `apt install openjdk-11-jre-headless cassandra` instead, _then_ it works. So there's something incorrect about using `openjdk-11-jdk` I guess?

File: integration_test/third_party_apps_data/applications/couchbase/metadata.yaml
Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/couchbase/metadata.yaml]: Removed `debian-11` from skipped platforms.

File: integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install
Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/mongodb/debian_ubuntu/install]: Removed all occurrences of `apt-get add`.

File: apps/mysql.go
Comment (jefferbrecht) [inline apps/mysql.go:325]:
```suggestion
					// When a timestamp is present, it is followed by a single tab character.
```

File: apps/mysql.go
Comment (jefferbrecht) [inline apps/mysql.go:346]: Does `timeRegexMariaDBNew` need to be included in here too? Or are those logs not multiline?

File: apps/mysql.go
Comment (jefferbrecht) [inline apps/mysql.go:607]: I assume it's not possible for `User@Host` to be the only line in an entire entry then? We'll always have a "next line" to consume?

File: integration_test/third_party_apps_data/applications/mariadb/debian_ubuntu/install
Comment (jefferbrecht) [inline integration_test/third_party_apps_data/applications/mariadb/debian_ubuntu/install:25]: Nit:

```suggestion
```

**Review by jefferbrecht — COMMENTED** [review]
LGTM, just minor comments.

We'll need to remove the do-not-submit from https://github.com/GoogleCloudPlatform/ops-agent/pull/1309/commits/dfb8c3f793ec30b785bbd3a0d6abd36a9f25a22f before merging, but let's wait for the last round of tests to finish first since I pushed some changes.

**Review by igorpeshansky — COMMENTED** [review]
Some preliminary comments…

File: integration_test/third_party_apps_data/applications/mongodb/centos_rhel/install
Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/mongodb/centos_rhel/install:3]: IIRC MongoDB 5 isn't available on Debian 11 ARM, so for simplicity I just upgraded all the tests to 6.0. This also unlocks testing on RHEL 9 which had no coverage before.

I don't want to keep piling more scope onto this PR so I'd rather not start a `mongodb5.0` test set here. If MongoDB 5 testing on RHEL is still valuable then perhaps we can special-case it so that 6.0 gets installed on RHEL 9 but 5.0 gets installed on other distros, as an interim measure while we work on a separate PR to add a `mongodb5.0` test set? WDYT?

Comment (igorpeshansky) [reply integration_test/third_party_apps_data/applications/mongodb/centos_rhel/install:3]: I'll leave it to you and ObservIQ (@dehaansa?) to determine how important it actually is to have 5.0 in our testing matrix. FWIW, I wasn't suggesting special-casing RHEL 9. It should be fairly straightforward to copy the old `mongodb` directory into `mongodb5.0` and exclude both `debian-11-arm64` and `rocky-linux-9` via `platforms_to_skip`, but we could do that in a separate PR if we decide it's worth it.

Comment (dehaansa) [reply integration_test/third_party_apps_data/applications/mongodb/centos_rhel/install:3]: Are there significant breaking changes between 5.0 and 6.0? If not, I don't see a requirement to continue to test 5.0, we aren't attempting to cover the full scope of supported versions in integration tests for other integrations, just versions where we encountered significant issues or know of breaking changes between versions. That said, if we want to add it as another supported integration to maintain support we can add 5.0 in the described way.

File: integration_test/testdata/prometheus/setup_json_exporter.sh
Comment (jefferbrecht) [reply integration_test/testdata/prometheus/setup_json_exporter.sh:12]: `dpkg` isn't portable though, so that simplicity would be offset by the complexity of finding equivalents on other distros...

The map version is nice because it's still portable yet simpler than what I had already, so I'll go with that for now unless there's an even simpler alternative that works on all distros.

File: integration_test/third_party_apps_data/applications/couchbase/debian_ubuntu/install
Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/couchbase/debian_ubuntu/install:9]: Yup, I'll change this to `${ARCH}`. And since this is the `debian_ubuntu` installer then I can use `$(dpkg --print-architecture)` too.

File: integration_test/testdata/prometheus/setup_json_exporter.sh
Comment (igorpeshansky) [reply integration_test/testdata/prometheus/setup_json_exporter.sh]: You're right, I'm so used to Debian using its own architecture constants that I didn't notice that this script is cross-distro.

File: integration_test/third_party_apps_data/applications/mongodb/centos_rhel/install
Comment (igorpeshansky) [reply integration_test/third_party_apps_data/applications/mongodb/centos_rhel/install:3]: Good point. I don't see anything in the [documentation](https://www.mongodb.com/docs/manual/release-notes/6.0-compatibility/) that could be potentially related to the way we collect metrics or logs, so maybe switching to 6.0 is not as big of a deal as I initially thought. Let's hold off on adding 5.0-specific testing until we discover breakage that wasn't caught by these.

Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/mongodb/centos_rhel/install:3]: SGTM 👍

File: confgenerator/logging_modify_fields.go
Comment (jefferbrecht) [reply confgenerator/logging_modify_fields.go:196]: No, the validation rules on `ModifyField.Type` prevent users from configuring this. It wouldn't be visible in any user-facing configurations.

Comment (igorpeshansky) [reply confgenerator/logging_modify_fields.go:196]: Want to add a comment to that effect, e.g.: `// internal-only type; external use blocked by field validation` or something?

File: integration_test/third_party_apps_data/applications/mariadb/debian_ubuntu/install
Comment (jefferbrecht) [reply integration_test/third_party_apps_data/applications/mariadb/debian_ubuntu/install:17]: Added a note to `metadata.yaml`.

File: apps/mysql.go
Comment (jefferbrecht) [reply apps/mysql.go:346]: Nevermind, looks like `timeRegexMariaDBNew` only applies to the error log.

File: integration_test/third_party_apps_test.go
Comment (igorpeshansky) [inline integration_test/third_party_apps_test.go:466]: I don't see `app` being used anywhere. Am I missing something?

Comment (jefferbrecht) [reply integration_test/third_party_apps_test.go:466]: Forgot to remove this in 478d040, will fix.

**Review by igorpeshansky — APPROVED** [review]
LGTM :shipit:

File: apps/mysql.go
Comment (quentinmit) [reply apps/mysql.go:607]: Correct, there's always at least one (and usually several) comment lines, and then at least one non-comment line before the next log line.

**Pull request merged**
